cli-plugins/manager/hooks.go

```diff
@@ -0,0 +1,140 @@
+package manager
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/docker/cli/cli/command"
+	"github.com/morikuni/aec"
+	"github.com/spf13/cobra"
+)
+
+// HookPluginData is the type representing the information
+// that plugins declaring support for hooks get passed when
+// being invoked following a CLI command execution
+type HookPluginData struct {
+	RootCmd string
+}
+
+// HookMessage represents a plugin hook response. Plugins
+// declaring support for CLI hooks need to print a json
+// representation of this type when their hook subcommand
+// is invoked.
+type HookMessage struct {
+	// TODO: should we have an enum to typify the kind of hook
+	// (such as `Next Steps`, `Warning`, etc.)?
+	Template string
+}
+
+// TemplateReplaceSubcommandName returns a hook template string
+// that will be replaced by the CLI subcommand being executed
+//
+// Example:
+//
+//	h := HookMessage{
+//	  Template: `you ran the subcommand: ` + TemplateReplaceSubcommandName(),
+//	}
+//
+// when being executed after the command:
+// `docker run --name "my-container" alpine`
+// will result in the message:
+// `you ran the subcommand: run`
+func TemplateReplaceSubcommandName() string {
+	return hookTemplateCommandName
+}
+
+// TemplateReplaceFlagValue returns a hook template string
+// that will be replaced by the flags value.
+//
+// Example:
+//
+//	h := HookMessage{
+//	  Template: `you ran a container named: ` + TemplateReplaceFlagValue("name"),
+//	}
+//
+// when being executed after the command:
+// `docker run --name "my-container" alpine`
+// will result in the message:
+// `you ran a container named: my-container`
+func TemplateReplaceFlagValue(flag string) string {
+	return fmt.Sprintf(hookTemplateFlagValue, flag)
+}
+
+// RunPluginHooks calls the hook subcommand for all present
+// CLI plugins that declare support for hooks in their metadata
+// and parses/prints their responses
+func RunPluginHooks(dockerCli command.Cli, cmd *cobra.Command, args []string) error {
+	plugins, err := ListPlugins(dockerCli, cmd)
```

Comment:
Wondering if this is something that's already done before we reach this (i.e., don't we scan for plugins on startup somewhere?).

Considering if we already have a list of plugins we know about, if we can re-use that list (assuming here that listing / scanning for plugins is potentially "heavy".

Reply:
Ah, hm. Had quick look, and it appears that we don't call `ListPlugins` in all cases, but only when (e.g.) doing `docker --help` (to create "stubs"), and `docker info` (to list plugins).

However, in those cases, we _would_ be running twice, but in other cases, we'd run `PluginRunCommand` (and/or `GetPlugin`); both of which also run `getPluginDirs` and `addPluginCandidatesFromDir`, and now I'm wondering if we need to memoize some of that (at least, we should look at benchmarking some of it).

Reply:
Yeah! Re: benchmarking, I worked up some general ones into the `#Performance Callouts` part of the PR description. For discussion/validation purposes, I think the implementation as-is works well, also given the rest of the comments in that section of the description – i.e., the performance penalty here only applies in interactive environments, etc.

However, I do think we should consider ways to improve performance here – and there, memoization might come in handy (or, as pointed out in that section, caching plugin metadata on disk.

Reply:
"resolving" this one since we've since had other conversations re: plugin metadata caching and socket communication.